monitor/monitor.go

```diff
@@ -77,21 +77,52 @@ func (m *Monitor) Run(npages int) {
 				break
 			}
 		}
-		if todo > 0 {
-			if err := monitorEvents.ReadAll(m.receiveEvent, m.lostEvent); err != nil {
-				log.WithError(err).Warn("Error received while reading from perf buffer")
+
+		if numListeners() > 0 {
+			if todo > 0 {
```

Comment:
epoll is used in its default edge-triggered mode here. Therefore, if you don't call `monitorEvents.ReadAll` every time `monitorEvents.Poll` returns `todo > 0`, the next call to `monitorEvents.Poll` will not block, because there's still data to be read from the the file descriptor. Therefore, this loop will spin and consume 100% of one CPU, and will also result in the perf ring buffer filling up.

In short, this `todo > 0` test CAN'T be conditioned on anything, e.g. `numListeners() > 0`.
You must revert this change.

You should either:
* Move your optimization into `m.receiveEvent`, where you could just ignore read events (don't decode them) if there are no listeners, as I previously suggested.
* Close the file descriptor until `numListeners() > 0`, as @tgraf suggested.

Reply:
Please wait till I request a new review. Thanks!